go/client/cmd_ctl_watchdog_windows.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package client
+
+import (
+	"bufio"
+	"errors"
+	"time"
+
+	"github.com/keybase/cli"
+	"github.com/keybase/client/go/libcmdline"
+	"github.com/keybase/client/go/libkb"
+	"github.com/natefinch/npipe"
+)
+
+type CmdWatchdog struct {
+	libkb.Contextified
+}
+
+func (c *CmdWatchdog) ParseArgv(ctx *cli.Context) error {
+	return nil
+}
+
+func (c *CmdWatchdog) Run() (err error) {
+	// Watch over the running service
+	// until it goes away, which will mean one of:
+	// - crash
+	// - system shutdown
+	// - uninstall
+	// - legitimate stoppage (ctl stop)
+
+	// Testing loop:
+	// - dial pipe. Pipe up equals service up.
+	// - Do a blocking read. Server won't be writing anything,
+	//   so this is our canary for shutdown.
+	//   - If down, test existence of PidFile (maybe wait a moment).
+	//      - No PidFile: normal shutdown. We stop too.
+	//      - PidFile still there: crashed(?) Restart.
+	//        (file should be writable in that case)
+	//
+	// Note that we give up after 10 consecutive crashes
+
+	if c.G().SocketInfo == nil {
+		return errors.New("Uninitialized socket")
+	}
+	pipeName := c.G().SocketInfo.GetFile()
+	if len(pipeName) == 0 {
+		return errors.New("No pipe name")
+	}
+
+	for i := 0; i < 10; i++ {
+		conn, err := npipe.DialTimeout(pipeName, time.Second*10)
+		if conn == nil {
+			// no service started. exit.
+			return err
+		}
+		for {
+			answer, err := bufio.NewReader(conn).ReadString('\n')
+			// We should not have received anything, this should mean
+			// the pipe has been closed - but test just in case
+			if len(answer) == 0 || err != nil {
+				break
+			}
+		}
+
+		conn.Close()
+
+		// Give the service a second to clean up its file
+		time.Sleep(time.Second * 1)
+
+		var fn string
+		if fn, err = c.G().Env.GetPidFile(); err != nil {
+			return err
+		}
+		crashed, _ := libkb.FileExists(fn)
+		if !crashed {
+			// apparently legitimate shutdown
+			return nil
+		}
+		// restart server case (is this the right command line?)
+		if err = ForkServer(c.G().Env.GetCommandLine(), c.G()); err != nil {
+			return err
+		}
+	}
+
+	return errors.New("Watchdog observed 10 crashes in a row. NOT reforking.")
```

Comment:
ah, i see why 10...  maybe still it would be nice to have a configuration for this